MultiBuild/BlueprintManager.cs

```diff
@@ -736,81 +646,201 @@ public static List<BuildPreview> Paste(Vector3 targetPos, float yaw, bool pasteI
                         bool overlappingBeltHasOutput = overlappingBelt.outputId != 0;
 
-                        if ((beltHasOutput && !overlappingBeltHasOutput) || (beltHasInput && !overlappingBeltHasInput ))
+                        if ((beltHasOutput && !overlappingBeltHasOutput) || (beltHasInput && !overlappingBeltHasInput))
                         {
                             // found overlapping belt that can be 'replaced' to connect to existing belts
                             preview.coverObjId = overlappingEntityId;
                             preview.ignoreCollider = true;
                             preview.willCover = true;
                         }
                     }
-                    
+
                 }
             }
 
-            
 
-            BuildLogic.ActivateColliders(ref actionBuild.nearcdLogic, absolutePositions);
+            return previews;
+        }
+
+        public static void PasteBuilding(int index, BuildingCopy building, Vector2 targetSpr, float yaw)
+        {
+
+            Vector2 newRelative = building.cursorRelativePos.Rotate(yaw * Mathf.Deg2Rad, building.originalSegmentCount);
+            Vector2 sprPos = newRelative + targetSpr;
+
+            float rawLatitudeIndex = (sprPos.x - Mathf.PI / 2) / 6.2831855f * 200;
+            int latitudeIndex = Mathf.FloorToInt(Mathf.Max(0f, Mathf.Abs(rawLatitudeIndex) - 0.1f));
+            int newSegmentCount = PlanetGrid.DetermineLongitudeSegmentCount(latitudeIndex, 200);
+
+            float sizeDeviation = building.originalSegmentCount / (float)newSegmentCount;
+
+            sprPos = new Vector2(newRelative.x, newRelative.y * sizeDeviation) + targetSpr;
 
-            if (!pasteInserters)
+            Vector3 absoluteBuildingPos = sprPos.SnapToGrid();
+
+            Quaternion absoluteBuildingRot = Maths.SphericalRotation(absoluteBuildingPos, yaw + building.cursorRelativeYaw);
+            PrefabDesc desc = GetPrefabDesc(building);
+            BuildPreview bp = BuildPreview.CreateSingle(building.itemProto, desc, true);
+            bp.ResetInfos();
+            bp.desc = desc;
+            bp.item = building.itemProto;
+            bp.recipeId = building.recipeId;
+            bp.lpos = absoluteBuildingPos;
+            bp.lrot = absoluteBuildingRot;
+
+            Pose pose = new Pose(absoluteBuildingPos, absoluteBuildingRot);
+
+            int objId = InserterPoses.AddOverride(pose, building.itemProto);
+
+            pastedEntities.Add(building.originalId, new PastedEntity()
             {
-                lastMaxWidth = currentMaxWidth;
-                return previews;
-            }
+                type = EPastedType.BUILDING,
+                index = index,
+                sourceBuilding = building,
+                pose = pose,
+                objId = objId,
+                buildPreview = bp
+            });
+
+            ActivateColliders(absoluteBuildingPos);
+            previews.Add(bp);
```

Comment:
You're accessing previews (shared static instance) without any kind of synchronization.

Reply:
ah ! right !
totally forgot about that !

Reply:
this method is not parallelised